File: tessoku/chapter3/a12.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, k int
	fmt.Fscan(in, &n, &k)

	var arr []int
	for i := 0; i < n; i++ {
		var a int
		fmt.Fscan(in, &a)
		arr = append(arr, a)
	}

	left := 0
	right := 1000000000

	for left < right {
		mid := (left + right) / 2

		answer := check(mid, n, k, arr)
		if answer {
			right = mid
		} else {
			left = mid + 1
		}
	}
	fmt.Fprintln(out, left)

}

func check(mid, n, k int, arr []int) bool {
	sum := 0
	for i := 0; i < n; i++ {
		sum += mid / arr[i]
	}
	if sum >= k {
		return true
	} else {
		return false
	}
}

File: tessoku/chapter3/a12.rb
# 答えがx以下 => true、xより大きい=> false
def check?(x, n, k, arr)
  sum = 0
  (0...n).each do |i|
    sum += x / arr[i]
  end

  sum >= k
end

n, k = gets.chomp.split(" ").map(&:to_i)
arr = gets.chomp.split(" ").map(&:to_i)

# 解について二部探索
# 解の範囲は1 ~ 1,000,000,000
# mid秒の結果はkより大きいか小さいか、で範囲を絞り込んでいく
left = 0
right = 1000000000

loop do
  break if left >= right

  mid = (left + right) / 2
  answer = check?(mid, n, k, arr)
  if answer
    right = mid
  else
    left = mid + 1
  end
end

puts left